### PR TITLE
Update how the start date and stop date of the TestRun is set Refs#23…

### PR DESCRIPTION
**Pull request by AirinB**
…23 :

- the start_date in the TestRun is not set automatically when the Test Run is created

- a toggle button is added for setting the start date and stop date, when the date is set for start or
  finish, the toggle disappears and only the timestamp is visible on the page

### CODE REVIEW

**Comment by AirinB** [comment]
@atodorov Here is a screenshot of how it's looking now. Seems to be working fine from what I have tested, but if the button is toggled many times and the refresh is pressed too fast, then the state could be inconsistent with what we could expect, not sure how I should handle that. Also, is valid the start_date to be set if the start date is not yet set? 
Screen Shot 2021-04-03 at 1 09 11 PM

File: tcms/testruns/models.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
 class TestRun(models.Model, UrlMixin):
     history = KiwiHistoricalRecords()
 
-    start_date = models.DateTimeField(auto_now_add=True, db_index=True)
+    start_date = models.DateTimeField(db_index=True, null=True, blank=True)
```
Comment (atodorov) [inline tcms/testruns/models.py:29]: See `sanity / test-for-missing-migrations` .  I believe this is what is causing the rest of the tests to fail as well.

File: tcms/testruns/templates/testruns/get.html [hunk 2]
```diff
@@ -63,39 +63,52 @@ <h2 class="card-pf-title" style="text-align: left">
...
                 <h2 class="card-pf-title" style="text-align: left">
                     <span class="fa fa fa-calendar-o"></span>{% trans 'Planned start' %}:
                     {{ object.planned_start|default:'-'}}
                 </h2>
 
-                <h2 class="card-pf-title" style="text-align: left">
-                    <span class="fa fa-calendar"></span>{% trans 'Started at' %}:
-                    {{ object.start_date }}
-                </h2>
+{#                Start date #}
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:72]: Comments are unnecessary b/c the code is explicit enough.

File: tcms/testruns/templates/testruns/get.html [hunk 3]
```diff
@@ -63,39 +63,52 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                {% if object.start_date == None %}
+                    <h2 class="card-pf-title" style="text-align: left">
+                        <span class="fa fa-spinner"></span>{% trans 'Start date' %}:
+                        <input class="bootstrap-switch" id="start_button" type="checkbox"
+
+                               {% if not perms.testruns.change_testrun %}disabled{% endif %}
+                               data-on-text="{% trans 'Started' %}" data-off-text="{% trans 'Planned' %}">
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:79]: When start_date is None => the button text should be "Start". The onclick() handler will update the date and also refresh the UI and hide itself. 

When start_date is set then you can't go back so the button will not be visible.


I don't think 2 state buttons make sense here b/c you have only 1 state. A simple button should do the trick, see https://www.patternfly.org/v3/pattern-library/widgets/#buttons + the Save buttons on most forms. Small or Extra Small should fit well into the existing UI. 

For a very minimal button you can use one of the checkmark icons + a title text like Start/Finish, etc: https://fontawesome.com/v4.7.0/icons/

File: tcms/testruns/templates/testruns/get.html [hunk 4]
```diff
@@ -63,39 +63,52 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                    </h2>
+                {% else %}
+                    <h2 class="card-pf-title" style="text-align: left">
+                        <span class="fa fa-calendar"></span>{% trans 'Started at' %}:
+                        <span class="start-date">
+                            {{ object.start_date|default:'-'}}
+                    </span>
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:86]: Don't repeat the file definition! Only surround the button with conditional template tags.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/templates/testruns/get.html (hunk 2 above)
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:72]: sure, I just place them for reference while I am still working not to get lost.

File: tcms/testruns/templates/testruns/get.html (hunk 3 above)
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:79]: should it be something like this or after the click the button should disappear?

Comment (atodorov) [reply tcms/testruns/templates/testruns/get.html:79]: Can't test until the problem with migrations is solved. See https://github.com/kiwitcms/Kiwi/pull/2328#discussion_r606699718

File: tcms/testruns/templates/testruns/get.html [hunk 5]
```diff
@@ -63,41 +63,41 @@ <h2 class="card-pf-title" style="text-align: left">
...
                     <span class="fa fa fa-calendar-o"></span>{% trans 'Planned start' %}:
                     {{ object.planned_start|default:'-'}}
                 </h2>
 
+{#                Start date #}
                 <h2 class="card-pf-title" style="text-align: left">
-                    <span class="fa fa-calendar"></span>{% trans 'Started at' %}:
-                    {{ object.start_date }}
+                    {% if object.start_date == None %}
+                        <span class="fa fa-check"></span>{% trans 'Start date' %}:
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:75]: Fix this comment first - https://github.com/kiwitcms/Kiwi/pull/2328#discussion_r606700870; then 

- fa-calendar stays as an icon for the field itself
- fa-check becomes the actual text/icon for the button and "Start" becomes a hint which shows on mouse hover.

File: tcms/testruns/templates/testruns/get.html [hunk 6]
```diff
@@ -63,41 +63,41 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                        <button class="btn btn-primary btn-xs" type="button" id="start-button">Start</button>
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:76]: Start - string not marked for translation.

File: tcms/testruns/static/testruns/js/get.js [hunk 7]
```diff
@@ -19,23 +19,26 @@ $(document).ready(() => {
...
-    })
+    $('#start-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'start_date': now }], testRun => {
+            const startDate = moment(testRun.start_date).format("DD MMM YYYY, HH:mm a")
+            $('.start-date').html(startDate)
+        })
+        });
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:30]: misaligned.

File: tcms/testruns/static/testruns/js/get.js [hunk 8]
```diff
@@ -19,23 +19,26 @@ $(document).ready(() => {
...
-            })
-        }
-    })
+    $('#start-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'start_date': now }], testRun => {
+            const startDate = moment(testRun.start_date).format("DD MMM YYYY, HH:mm a")
+            $('.start-date').html(startDate)
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:28]: button should hide itself after updating the field.

File: tcms/testruns/static/testruns/js/get.js [hunk 9]
```diff
@@ -19,23 +19,26 @@ $(document).ready(() => {
...
+        })
+        });
+
+    $('#stop-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'stop_date': now }], testRun => {
+            const stopDate = moment(testRun.stop_date).format("DD MMM YYYY, HH:mm a")
+            $('.stop-date').html(stopDate)
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:38]: button should hide itself and also update page title, see the deleted code.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by AirinB** [comment]
@atodorov here is how it is looking right now and the button disappears after it is clicked. I think I should work a bit at the design, but let me know if the functionality it is okay right now :D
Screen Shot 2021-04-06 at 5 29 36 PM

**Review by atodorov — CHANGES_REQUESTED** [review]
Buttons functionality looks good, other changes requested.

Note: don't mark review comments as resolved b/c I have to expand all of them to actually check which ones have been resolved. That is the job of the reviewer.

File: tcms/testruns/templates/testruns/get.html [hunk 10]
```diff
@@ -63,41 +63,41 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                            <i class="fa fa-check"></i>
+                        </button>
+                    {% endif %}
                     <span class="stop-date">
-                        {% if object.stop_date %}
-                            {{ object.stop_date }}
-                        {% else %}
-                            -
-                        {% endif %}
+                        {{ object.stop_date|default:'' }}
```
[the remark this answers is not in the document]
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:98]: it's just that this line will be visible when the date is not set yet

**Comment by AirinB** [comment]
@atodorov current UI

File: tcms/testruns/templates/testruns/get.html (hunk 10 above)
[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/testruns/templates/testruns/get.html]: Yes, but all of the others are the same so they need to be consistent.

File: tcms/testruns/templates/testruns/get.html [hunk 11]
```diff
@@ -78,26 +71,33 @@ <h2 class="card-pf-title" style="text-align: left">
...
                     <span class="fa fa fa-calendar-o"></span>{% trans 'Planned stop' %}:
                     {{ object.planned_stop|default:'-' }}
                 </h2>
-
                 <h2 class="card-pf-title" style="text-align: left">
                     <span class="fa fa-calendar-check-o"></span>{% trans 'Finished at' %}:
+                    {% if object.stop_date == None %}
+                        <button class="btn btn-warning btn-xs" type="button" id="stop-button"
+                                data-toggle="tooltip" data-placement="top" title="{% trans 'Stop' %}">
+                            <i class="fa fa-check" style="color: white; margin: auto; border-radius: 30px; "></i>
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:94]: `border-radius` not needed IMO. I don't see any difference with or without it.

File: tcms/testruns/templates/testruns/get.html [hunk 12]
```diff
@@ -78,26 +71,33 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                        </button>
+                    {% endif %}
                     <span class="stop-date">
-                        {% if object.stop_date %}
-                            {{ object.stop_date }}
-                        {% else %}
-                            -
-                        {% endif %}
+                        {{ object.stop_date|default:'-' }}
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:98]: If you move the field value before the button it looks better when the default is `-`

File: tcms/testruns/templates/testruns/get.html [hunk 13]
```diff
@@ -78,26 +71,33 @@ <h2 class="card-pf-title" style="text-align: left">
...
 
+                </h2>
                 <h2 class="card-pf-title" style="text-align: left">
                     <span class="fa fa fa-calendar-o"></span>{% trans 'Planned stop' %}:
                     {{ object.planned_stop|default:'-' }}
                 </h2>
-
                 <h2 class="card-pf-title" style="text-align: left">
                     <span class="fa fa-calendar-check-o"></span>{% trans 'Finished at' %}:
+                    {% if object.stop_date == None %}
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:91]: The canonical version is `if not object.stop_date` b/c None and other zero values are evaluated as False in boolean context.

File: tcms/testruns/templates/testruns/get.html [hunk 14]
```diff
@@ -78,26 +71,33 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                        <button class="btn btn-warning btn-xs" type="button" id="stop-button"
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:92]: I tried moving this button to the right with `float:right` and it looks a bit better but then the card height gets shorter and the button  goes outside of the card. If you can figure out how to make it look presentable I'd prefer the buttons to the right. If not leave it as is.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/static/testruns/js/get.js [hunk 15]
```diff
@@ -9,6 +9,9 @@ const allExecutionStatuses = {},
     autocomplete_cache = {}
 
 $(document).ready(() => {
+    $(function () {
+        $('[data-toggle="tooltip"]').tooltip()
+    })
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:14]: Let's remove this block. Regular tooltips are fine. 

If you insist on keeping the nicer tooltips the onclick handler of the buttons should hide them as well. Try these steps:
1) hover the mouse on the Stop button => tooltip appears
2) click the Stop button without moving the mouse => timestamp is updated; button hides itself

however the tooltip stays on the screen.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/templates/testruns/get.html (hunk 11 above)
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:94]: yes, it doesn't have any effect. I was just experimenting.

File: tcms/testruns/templates/testruns/get.html (hunk 12 above)
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:98]: true! Thank you!

File: tcms/testruns/static/testruns/js/get.js (hunk 15 above)
Comment (AirinB) [reply tcms/testruns/static/testruns/js/get.js:14]: @atodorov basically I tried and on to reproduce this problem and it didn't really appear in my case (MacBook pro Chrome and Safari), but I have no hard feeling about it. I could delete it if you think it's not needed

**Comment by AirinB** [comment]
@atodorov the UI
Screen Shot 2021-04-12 at 4 22 22 PM

File: tcms/testruns/static/testruns/js/get.js (hunk 15 above)
Comment (AirinB) [reply tcms/testruns/static/testruns/js/get.js]: I updated this function to explicitly call the trigger the tooltip when the mouse hovers only, this was one of the solutions on stack overflow, but not sure if it has an effect on your side.

File: tcms/testruns/static/testruns/js/get.js (hunk 9 above)
Comment (atodorov) [reply tcms/testruns/static/testruns/js/get.js]: Still unresolved.

File: tcms/testruns/static/testruns/js/get.js [hunk 16]
```diff
@@ -19,23 +25,30 @@ $(document).ready(() => {
...
+    $('#start-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'start_date': now }], testRun => {
+            const startDate = moment(testRun.start_date).format("DD MMM YYYY, HH:mm a")
+            $('.start-date').html(startDate);
+            $(this).hide();
+        })
+
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:37]: Remove empty line.

File: tcms/testruns/static/testruns/js/get.js [hunk 17]
```diff
@@ -19,23 +25,30 @@ $(document).ready(() => {
...
+    $('#stop-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'stop_date': now }], testRun => {
+            const stopDate = moment(testRun.stop_date).format("DD MMM YYYY, HH:mm a")
+            $('.stop-date').html(stopDate);
+            $(this).hide();
+        })
+
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:49]: Remove empty line.

File: tcms/testruns/static/testruns/js/get.js [hunk 18]
```diff
@@ -19,23 +25,30 @@ $(document).ready(() => {
...
+    });
+
+    $('#stop-button').on('click', function () {
+        const timeZone = $('#clock').data('time-zone')
+        const now = currentTimeWithTimezone(timeZone)
+
+        jsonRPC('TestRun.update', [testRunId, { 'stop_date': now }], testRun => {
+            const stopDate = moment(testRun.stop_date).format("DD MMM YYYY, HH:mm a")
+            $('.stop-date').html(stopDate);
+            $(this).hide();
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:47]: Still not updating title like it used to be. See https://github.com/kiwitcms/Kiwi/pull/2328#discussion_r607568638

File: tcms/testruns/templates/testruns/get.html [hunk 19]
```diff
@@ -63,41 +63,44 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                <h2 class="card-pf-title" style="text-align: left;">
+                    <span class="fa fa-calendar"></span>
+                        {% trans 'Started at' %}:
+                    <span class="start-date">
+                            {{ object.start_date|default:'-' }}
+                        </span>
+                    {% if not object.start_date %}
+                        <button class="btn btn-success btn-xs" type="button" id="start-button"
+                                data-toggle="tooltip" data-placement="top" title="{% trans 'Start' %}"
+                        style="float:right;">
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:81]: can you either 
1) align the attribites by column so they are more readable or
2) move this at the end of previous line like you have on L98-99.

File: tcms/testruns/templates/testruns/get.html [hunk 20]
```diff
@@ -63,41 +63,44 @@ <h2 class="card-pf-title" style="text-align: left">
...
                     <span class="fa fa fa-calendar-o"></span>{% trans 'Planned start' %}:
                     {{ object.planned_start|default:'-'}}
                 </h2>
 
-                <h2 class="card-pf-title" style="text-align: left">
-                    <span class="fa fa-calendar"></span>{% trans 'Started at' %}:
-                    {{ object.start_date }}
+                <h2 class="card-pf-title" style="text-align: left;">
+                    <span class="fa fa-calendar"></span>
+                        {% trans 'Started at' %}:
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:74]: Is there a reason for these 3 lines to be updated? Apart from a trailing `;` and changing the line numbers I don't see a functional change which warrants this.

File: tcms/testruns/static/testruns/js/get.js [hunk 21]
```diff
@@ -10,6 +10,12 @@ const allExecutionStatuses = {},
 
 $(document).ready(() => {
 
+    $('[data-toggle="tooltip"]').tooltip({
+        trigger: 'hover'
+    });
+
+
+
```
Comment (atodorov) [inline tcms/testruns/static/testruns/js/get.js:18]: Leave only 1 empty line.

**Review by atodorov — CHANGES_REQUESTED** [review]
Few changes requested

- still missing one functional change to match the old behavior
- the rest are cosmetic issues
- Commit log reference should be updated to `Fixes` b/c once we merge this PR the issue will be fixed.

**Review by atodorov — CHANGES_REQUESTED** [review]
Almost there, see `lint / black` failures.

**Comment by AirinB** [comment]
> Almost there, see `lint / black` failures.

Done

File: tcms/testruns/templates/testruns/get.html [hunk 22]
```diff
@@ -79,25 +72,35 @@ <h2 class="card-pf-title" style="text-align: left">
...
                     </span>
+                    {% if not object.stop_date %}
+                        <button class="btn btn-warning btn-xs pull-right" type="button"        id="stop-button"
+                                data-toggle="tooltip"                     data-placement="top" title="{% trans 'Stop' %}" style="margin-bottom: 1rem;">
+                            <i class="fa fa-check" style="color: white; margin: auto; "></i>
+                        </button>
+                    {% endif %}
                 </h2>
-
-                <div class="card-pf-body"></div>
```
Comment (atodorov) [inline tcms/testruns/templates/testruns/get.html:100]: This is a mandatory element, eventhough it is empty. Bring it back.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/templates/testruns/get.html (hunk 22 above)
Comment (AirinB) [reply tcms/testruns/templates/testruns/get.html:100]: @atodorov done

**Pull request merged**